compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3135,36 +3122,23 @@ generateS390CompareAndBranchOpsHelper(TR::Node * node, TR::CodeGenerator * cg, T
    else
       {
       TR_S390BinaryCommutativeAnalyser temp(cg);
-
-      // FIXME: Why are we doing this business with the secondChild's type
-      // here? How can the second child's type ever be different from the
-      // first child's type?
-      // Answer:The only case where it can be different is in the compressed refs
-      // build. an aload of a class pointer is 32-bits only.
-      //
-      TR::DataType childType = secondChild->getDataType();
-      if (childType == TR::Address)
+      TR_ASSERT_FATAL(firstChild->getDataType() == secondChild->getDataType(), "Data types of both operands should match for generating compare and branch.");
+      // On 64-Bit platform with compressed references, it is possible that one of the child is loading a class from the object while other can be
+      // loading it from the temp stack slot in this case, VFT load should be evaluated into register where it loads 4 bytes with lowest byte cleared and
+      // zero extend it so we will have a VFT pointer in 64-Bit register so when it should generate instruction to compare 8 bytes.
+      TR::DataType childType = TR::Compiler->target.is32Bit() && firstChild->getDataType() == TR::Address ? TR::Int32 : firstChild->getDataType();
+      if (TR::Compiler->target.is64Bit())
```

Comment:
Fix looks good. The comment has a few grammatical issues. I've also further elaborated on the problem and made the assert better by displaying the actual data types encountered as that is something the user would want to know if the assert is ever hit:

```
      TR_ASSERT_FATAL(firstChild->getDataType() == secondChild->getDataType(), "Data type of firstChild (%s) and secondChild (%s) should match for generating compare and branch", firstChild->getDataType().toString(), secondChild->getDataType().toString());

      // On 64-Bit platforms with compressed references it is possible that one (or both) of the children of the 
      // compare is loading a class from the object (VFT symbol). This symbol is specially treated within the JIT at
      // the moment because it is an `aloadi` which loads a 32-bit value (64-bit compressed references) and zero
      // extends it to a 64-bit address. Unfortunately the generic analyzers used below are unaware of this fact
      // and could end up generating an incorrect instruction to evaluate such a VFT load.
      //
      // To prevent such logic leaking into the generic analyzers (which should remain generic) we force the evaluation
      // of such VFT nodes here before calling out to the analyzer.
      TR::DataType childType = TR::Compiler->target.is32Bit() && firstChild->getDataType() == TR::Address ? TR::Int32 : firstChild->getDataType();
      if (TR::Compiler->target.is64Bit())
```

Reply:
Changed in https://github.com/eclipse/omr/commit/72d58c3cd5edcd89e97671d490393e1b8e01f9db